src/UI/Button.elm

```diff
@@ -32,283 +31,502 @@ import Element.Border as Border
 import UI.Internal.Palette as Palette
 import UI.Internal.Primitives as Primitives
 import UI.Internal.Size as Size exposing (Size)
 import UI.Internal.Text as Text exposing (TextColor)
 import UI.Link as Link exposing (Link)
-import UI.Palette as Palette exposing (brightnessDarkest, brightnessLight, brightnessLighter, brightnessLightest, brightnessMiddle)
+import UI.Palette as Palette exposing (brightnessDarkest, brightnessLight, brightnessLighter, brightnessLightest, brightnessMiddle, tonePrimary)
 import UI.RenderConfig exposing (RenderConfig)
 import UI.Text as Text
 import UI.Utils.ARIA as ARIA
 import UI.Utils.Element as Element
 
 
 type alias Options =
-    { mode : ButtonMode
-    , tone : ButtonTone
-    , width : ButtonWidth
+    { width : ButtonWidth
     , size : Size
     }
 
 
 type alias Properties msg =
     { body : ButtonBody
-    , click : ButtonClick msg
+    , mode : ButtonMode msg
     }
 
 
 type Button msg
     = Button (Properties msg) Options
+    | Toggle (ToggleProperties msg) Options
 
 
 type ButtonBody
     = BodyText String
     | BodyIcon Icon
 
 
-type ButtonClick msg
-    = ClickMsg msg
-    | ClickLink Link
-    | ClickToggle (Bool -> msg) Bool
+type ButtonAction msg
+    = ActionMsg msg
+    | ActionRedirect Link
 
 
-type ButtonMode
-    = ModeEnabled
-    | ModeDisabled
+type ButtonMode msg
+    = ButtonWorking (ButtonAction msg) ButtonStyle
+    | ButtonDisabled
+    | ButtonSuccess
 
 
-type ButtonTone
-    = ToneDanger
+type EmbossedTone
+    = TonePrimary
+    | ToneDanger
     | ToneLight
-    | TonePrimary
-    | ToneSuccess
     | ToneClear
 
 
+type ButtonStyle
+    = StyleEmbossed EmbossedTone
+    | StyleHyperlink
+
+
 type ButtonWidth
     = WidthFull
-    | WidthRelative
+    | WidthShrink
+
+
+type alias ToggleProperties msg =
+    { current : Bool
+    , toggleMsg : Bool -> msg
+    , hint : String
+    }
 
 
 
--- Create
+-- Init
 
 
 defaultOptions : Options
 defaultOptions =
-    { mode = ModeEnabled
-    , tone = TonePrimary
-    , width = WidthRelative
+    { width = WidthShrink
     , size = Size.default
     }
 
 
-buttonAny : ButtonClick msg -> ButtonBody -> Button msg
-buttonAny click body =
-    Button (Properties body click) defaultOptions
+toggle : String -> (Bool -> msg) -> Bool -> Button msg
+toggle hint msg isEnabled =
+    Toggle { toggleMsg = msg, current = isEnabled, hint = hint } defaultOptions
 
 
-button : msg -> ButtonBody -> Button msg
-button msg =
-    buttonAny (ClickMsg msg)
+fromLabel : String -> ButtonBody
+fromLabel label =
+    BodyText label
 
 
-toggle : (Bool -> msg) -> Bool -> ButtonBody -> Button msg
-toggle msg isEnabled =
-    buttonAny (ClickToggle msg isEnabled)
+fromIcon : Icon -> ButtonBody
+fromIcon icon =
+    BodyIcon icon
 
 
-link : Link -> ButtonBody -> Button msg
-link realLink =
-    buttonAny (ClickLink realLink)
 
+-- Body to Button
 
 
--- Options
+disabled : ButtonBody -> Button msg
```

Comment:
So, maybe we don't need to have disabled and success as creation functions, I think those 2 states are always conditionally linked to something. For example. I want to disable the button when the text field is empty. I don't see a case that someone is using those functions instead of `withDisableIf` or `withSuccessIf`.